File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
class Solution {
public:  
    
    bool isValidBST(TreeNode* root) {
        stack<TreeNode*> st;
        TreeNode* prev = NULL;
        while( root || !st.empty()){
            while(root){
                st.push(root);
                root = root->left;
            }
            TreeNode* popped = st.top();
            st.pop();
            
            if(prev && prev->val >= popped->val) return false;
            prev = popped;
            root = popped->right;
        }
        return true;
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root, minm = float("-inf"), maxm = float("inf")):
            return (not root) or (minm <root.val<maxm and (self.isValidBST(root.left, minm, root.val)) and (self.isValidBST(root.right, root.val, maxm)))